Mathlib/Analysis/SpecialFunctions/BinaryEntropy.lean

```diff
@@ -0,0 +1,428 @@
+  _ = log 2 := by rw [mul_inv_cancel₀, mul_inv_cancel₀, one_add_one_eq_two] <;> positivity
+
+lemma binEntropy_le_log_two : binEntropy p ≤ log 2 := by
+  obtain rfl | hp := eq_or_ne p 2⁻¹
+  · simp
+  · exact (binEntropy_lt_log_two.2 hp).le
+
+lemma binEntropy_eq_log_two : binEntropy p = log 2 ↔ p = 2⁻¹ := by
+  rw [binEntropy_le_log_two.eq_iff_not_lt, binEntropy_lt_log_two, not_ne_iff]
+
+/-- Binary entropy is continuous everywhere.
+This is due to definition of `Real.log` for negative numbers. -/
+@[fun_prop] lemma binEntropy_continuous : Continuous binEntropy := by
+  rw [binEntropy_eq_negMulLog_add_negMulLog_one_sub']; fun_prop
+
+@[fun_prop] lemma differentiableAt_binEntropy (hp₀ : p ≠ 0) (hp₁ : p ≠ 1) :
+    DifferentiableAt ℝ binEntropy p := by
+  rw [ne_comm, ← sub_ne_zero] at hp₁
+  unfold binEntropy
+  simp only [log_inv, mul_neg]
+  fun_prop (disch := assumption)
+
+set_option push_neg.use_distrib true in
+lemma differentiableAt_binEntropy_iff_ne_zero_one :
+    DifferentiableAt ℝ binEntropy p ↔ p ≠ 0 ∧ p ≠ 1 := by
+  refine ⟨fun h ↦ ⟨?_, ?_⟩, fun h ↦ differentiableAt_binEntropy h.1 h.2⟩
+    <;> rintro rfl <;> unfold binEntropy at h
+  · rw [DifferentiableAt.add_iff_left] at h
+    simp [log_inv, mul_neg, ← neg_mul, ← negMulLog_def, differentiableAt_negMulLog_iff] at h
+    fun_prop (disch := simp)
+  · rw [DifferentiableAt.add_iff_right, differentiableAt_iff_comp_const_sub (b := 1)] at h
+    simp [log_inv, mul_neg, ← neg_mul, ← negMulLog_def, differentiableAt_negMulLog_iff] at h
+    fun_prop (disch := simp)
+
+set_option push_neg.use_distrib true in
+/-- Binary entropy has derivative `log (1 - p) - log p`.
+It's not differentiable at `0` or `1` but the junk values of `deriv` and `log` coincide there. -/
+lemma deriv_binEntropy (p : ℝ) : deriv binEntropy p = log (1 - p) - log p := by
+  by_cases hp : p ≠ 0 ∧ p ≠ 1
+  · obtain ⟨hp₀, hp₁⟩ := hp
+    rw [ne_comm, ← sub_ne_zero] at hp₁
+    rw [binEntropy_eq_negMulLog_add_negMulLog_one_sub', deriv_add, deriv_comp_const_sub,
+      deriv_negMulLog hp₀, deriv_negMulLog hp₁]
+    ring
+    all_goals fun_prop (disch := assumption)
+  -- pathological case where `deriv = 0` since `binEntropy` is not differentiable there
+  · rw [deriv_zero_of_not_differentiableAt (differentiableAt_binEntropy_iff_ne_zero_one.not.2 hp)]
+    push_neg at hp
+    obtain rfl | rfl := hp <;> simp
+
+/-! ### `q`-ary entropy -/
+
+/-- Shannon q-ary Entropy function (measured in Nats, i.e., using natural logs).
+
+It's the Shannon entropy of a random variable with possible outcomes {1, ..., q}
+where outcome `1` has probability `1 - p` and all other outcomes are equally likely.
+
+The usual domain of definition is p ∈ [0,1], i.e., input is a probability.
+
+This is a generalization of the binary entropy function `binEntropy`. -/
+@[pp_nodot] noncomputable def qaryEntropy (q : ℕ) (p : ℝ) : ℝ := p * log (q - 1 : ℤ) + binEntropy p
```

Comment:
If at all possible, I think it would be very nice if the name could avoid hard-coding a variable name. I'm still sad that we have the name `padicInt` in mathlib, and the `q` in `qary` is another example.

Could this be called `shannonEntropy` instead? Or would `aryEntropy` be an option?

Anyway, I don't want to block this PR on this issue, if you think another name doesn't work well...

Reply:
I don't think `shannonEntropy` is a good name, as this is merely a special case of the Shannon entropy, namely it is the Shannon entropy of a random variable on `q` values, where one value is taken with probability `1 - p` and the other `q - 1` values are equally likely.

Reply:
I believe "q-ary entropy" is *the* standard name for this concept. It was added upon suggestion by @dupuisf , what do you think? 

Shannon entropy is for sure a distinct and more general concept that will be in Mathlib as well (https://github.com/leanprover-community/mathlib4/pull/10493 although it seems stalled for now).

Reply:
Yeah, `shannonEntropy` suggests something very different (and we want to keep that name available for that). Unfortunately I can't think of another name for this, and certainly not one that people will spontaneously search for.

Reply:
@jcommelin I think that means we would like to keep the name.